src/com/amazon/corretto/crypto/provider/AesGcmSpi.java

```diff
@@ -739,6 +643,153 @@ protected synchronized int engineDoFinal(byte[] bytes, int offset, int length, b
+            if (context != null) {
+                // We already have a context, so let's reuse it.
+                outLen = context.use(ptr -> {
+                    return oneShotDecrypt(
+                            ptr,
+                            null,
+                            workingInputArray,
+                            workingInputOffset,
+                            workingInputLength,
+
+                            result,
+                            0,
+
+                            tagLength,
+                            key,
+                            iv,
+
+                            // The cost of calling decryptAADBuf.getDataBuffer() when its buffer is empty is significant for 16-byte
+                            // decrypt operations (approximately a 7% performance hit). To avoid this, we reuse the same empty array
+                            // instead in this common-case path.
+                            decryptAADBuf.size() != 0 ? decryptAADBuf.getDataBuffer() : EMPTY_ARRAY,
+                            decryptAADBuf.size()
+                    );
+                });
+            } else {
+                // We don't have an existing context, however we might want to save one
+                final long[] ptrOut = keyUsageCount > KEY_REUSE_THRESHOLD ? new long[1] : null;
+                outLen = oneShotDecrypt(
+                        0,
+                        ptrOut,
+                        workingInputArray,
+                        workingInputOffset,
+                        workingInputLength,
+
+                        result,
+                        0,
+
+                        tagLength,
+                        key,
+                        iv,
+
+                        // The cost of calling decryptAADBuf.getDataBuffer() when its buffer is empty is significant for 16-byte
+                        // decrypt operations (approximately a 7% performance hit). To avoid this, we reuse the same empty array
+                        // instead in this common-case path.
+                        decryptAADBuf.size() != 0 ? decryptAADBuf.getDataBuffer() : EMPTY_ARRAY,
+                        decryptAADBuf.size()
+                );
+
+                if (ptrOut != null) {
+                    context = new NativeContext(ptrOut[0]);
+                }
+            }
+            // Decryption completed successfully.
+            if (outLen == result.length) {
+                // This should always be the case.
+                return result;
+            } else if (outLen < result.length) {
+                // While technically allowed by the APIs, this should never happen.
+                return Arrays.copyOf(result, outLen);
+            } else {
+                throw new RuntimeCryptoException("Output longer than expected.");
+            }
+        } finally {
+            stateReset();
+        }
+    }
+
+    @Override
+    protected byte[] engineDoFinal(byte[] bytes, int offset, int length)
```

Comment:
I'll add synchronization. Thank you.

We do support multi-shot here, or am I missing something?

Reply:
> We do support multi-shot here, or am I missing something?

nope, that's my understanding as well. i was calling that out this could be used multi-shot in case you had skipped the `synchronized` qualifier thinking this would only be called in a single-shot manner (and thus not requiring serialization with calls to `engineUpdate`).